packages/api/src/folder/service.ts

```diff
@@ -0,0 +1,122 @@
+import createError from "http-errors";
+import { logger } from "@stela/logger";
+import { db } from "../database";
+import type {
+  Folder,
+  FolderColumnsForUpdate,
+  FolderLink,
+  PatchFolderRequest,
+} from "./models";
+import { requestFieldsToDatabaseFields } from "./helper";
+import type { GetAccountArchiveResult } from "../account/models";
+import { permission } from "../access/permission";
+import { folderAccess } from "./access";
+
+export const getFolderById = async (requestQuery: {
+  folderId: string;
+}): Promise<Folder> => {
+  const folder = await db
+    .sql<Folder>("folder.queries.get_folder_by_id", {
+      folderId: requestQuery.folderId,
+    })
+    .catch((err) => {
+      logger.error(err);
+      throw new createError.InternalServerError("Failed to retrieve folder");
+    });
+
+  if (folder.rows[0] === undefined) {
+    throw new createError.NotFound("Folder not found.");
+  }
+
+  return folder.rows[0];
+};
+
+export const getFolderLinkByFolderId = async (requestQuery: {
+  folderId: string;
+}): Promise<FolderLink> => {
+  const folder = await db
+    .sql<FolderLink>("folder.queries.get_folder_link_by_folder_id", {
+      folderId: requestQuery.folderId,
+    })
+    .catch((err) => {
+      logger.error(err);
+      throw new createError.InternalServerError(
+        "Failed to retrieve folder link"
+      );
+    });
+
+  if (folder.rows[0] === undefined) {
+    throw new createError.NotFound("Folder link not found.");
+  }
+
+  return folder.rows[0];
+};
+
+function buildPatchQuery(patchFolderRequest: FolderColumnsForUpdate): string {
+  const updates = Object.entries(patchFolderRequest)
+    .filter(([key, value]) => value !== undefined && key !== "folderId")
+    .map(([key, _]) => `${key} = :${key}`);
+
+  if (updates.length === 0) {
+    throw new createError.BadRequest("Request cannot be empty");
+  }
+
+  const query = `
+    UPDATE folder
+    SET ${updates.join(", ")}
+    WHERE folderid = :folderId
+    RETURNING folder.folderid AS "folderId"
+  `;
+
+  return query.trim();
+}
+
+const validateCanPatchFolder = async (
+  folderId: string,
+  emailFromAuthToken: string
+): Promise<void> => {
+  const folder = await getFolderById({
+    folderId,
+  });
+  const accountArchiveResult = await db
+    .sql<GetAccountArchiveResult>("account.queries.get_account_archive", {
+      archiveId: folder.archiveId,
+      email: emailFromAuthToken,
+    })
+    .catch((err) => {
+      logger.error(err);
+      throw new createError.InternalServerError(
+        "Failed to retrieve account archive"
+      );
+    });
+  const accountArchive = accountArchiveResult.rows[0];
```

Comment:
If the permissions come by way of an `account_archive` record, we need to check that it has edit-level permissions. Note that the Manager role exists on archive memberships, but does not exist on shares (and therefore does not exist on `access` records)